src/hotspot/os/linux/hugepages.cpp

```diff
@@ -0,0 +1,236 @@
+  // for a line with the following format:
+  //    Hugepagesize:     2048 kB
+  //
+  // If we can't determine the value (e.g. /proc is not mounted, or the text
+  // format has been changed), we'll set largest page size to 0
+
+  FILE *fp = os::fopen("/proc/meminfo", "r");
+  if (fp) {
+    while (!feof(fp)) {
+      int x = 0;
+      char buf[16];
+      if (fscanf(fp, "Hugepagesize: %d", &x) == 1) {
+        if (x && fgets(buf, sizeof(buf), fp) && strcmp(buf, " kB\n") == 0) {
+          pagesize = x * K;
+          break;
+        }
+      } else {
+        // skip to next line
+        for (;;) {
+          int ch = fgetc(fp);
+          if (ch == EOF || ch == (int)'\n') break;
+        }
+      }
+    }
+    fclose(fp);
+  }
+
+  return pagesize;
+}
+
+// Given a file that contains a single (integral) number, return that number, 0 and false if failed.
+static bool read_number_file(const char* file, size_t* out) {
+  (*out) = 0;
+  FILE* f = ::fopen(file, "r");
+  bool rc = false;
+  if (f != nullptr) {
+    uint64_t i = 0;
+    if (::fscanf(f, SIZE_FORMAT, out) == 1) {
+      rc = true;
+    }
+    ::fclose(f);
+  }
+  return rc;
+}
+
+static const char* const sys_hugepages = "/sys/kernel/mm/hugepages";
+
+// Scan all directories in /sys/kernel/mm/hugepages/hugepages-xxxx
+// to discover the available page sizes
+static os::PageSizes scan_hugepages() {
+
+  os::PageSizes pagesizes;
+
+  DIR *dir = opendir(sys_hugepages);
+
+  struct dirent *entry;
+  size_t pagesize;
+  while ((entry = readdir(dir)) != nullptr) {
+    if (entry->d_type == DT_DIR &&
+        sscanf(entry->d_name, "hugepages-%zukB", &pagesize) == 1) {
+      // The kernel is using kB, hotspot uses bytes
+      // Add each found Large Page Size to page_sizes
+      pagesize *= K;
+      pagesizes.add(pagesize);
+    }
+  }
+  closedir(dir);
+
+  return pagesizes;
+}
+
+void StaticHugePageSupport::print_on(outputStream* os) {
+  if (_initialized) {
+    os->print("Static hugepage support: ");
+    bool first = true;
+    for (size_t s = _pagesizes.smallest(); s != 0; s = _pagesizes.next_larger(s)) {
+      os->print("%s" EXACTFMT, (first ? "" : ", "), EXACTFMTARGS(s));
+      if (_default_hugepage_size == s) {
+        os->print_cr(" (default)");
+      }
+      first = false;
+    }
+    os->print_cr("  default pagesize: " EXACTFMT, EXACTFMTARGS(_default_hugepage_size));
+  } else {
+    os->print_cr("  unknown.");
+  }
+}
+
+void StaticHugePageSupport::scan_os() {
+  _pagesizes = scan_hugepages();
+  _default_hugepage_size = scan_default_hugepagesize();
+  assert(_pagesizes.contains(_default_hugepage_size),
+         "Unexpected configuration: default pagesize (" SIZE_FORMAT ") "
+         "has no associated directory in /sys/kernel/mm/hugepages..", _default_hugepage_size);
+  _initialized = true;
+  LogTarget(Info, pagesize) lt;
+  if (lt.is_enabled()) {
+    LogStream ls(lt);
+    print_on(&ls);
+  }
```

Comment:
As `print_on` prints several lines you want to use `NonInterleavingLogStream`.

Do this:

```c++
LogMessage(pagesize) lm;
NonInterleavingLogStream ls{LogLevelType::Info, lm};
if (ls.is_enabled()) {
  print_on(&ls);
}
```

Reply:
I'm just using `{}`-init here, it's the same as using `()` so your choice on which to use.

Reply:
Okay. We are at initialization time, though; everything should still be single-threaded here. I dislike providing the log level twice, but LogTarget does not expose it.

Reply:
Wait, if I do this, I need to backport [8286180](https://bugs.openjdk.org/browse/JDK-8286180) before backporting this fix., right? :(

Oh well, chances are this does not integrate cleanly anyway.

Reply:
The more I think about this, the more doubts I have, sorry. Logging should not cost anything if disabled. But with your proposal, I pay for the construction of both LogMessage and NonInterleavingLogStream every time.

Reply:
>But with your proposal, I pay for the construction of both LogMessage and NonInterleavingLogStream every time.

Yeah, I made a mental footnote that `LogTarget` should be able to be used to initialize these inside of an if-block. UL is still a bit painful about these things. My suggestion sounds like a massive pain, I'm happy if you don't do it :).

Reply:
> > But with your proposal, I pay for the construction of both LogMessage and NonInterleavingLogStream every time.
> 
> Yeah, I made a mental footnote that `LogTarget` should be able to be used to initialize these inside of an if-block. UL is still a bit painful about these things. My suggestion sounds like a massive pain, I'm happy if you don't do it :).

Thank you @jdksjolen !